### PR TITLE
fix #887: New lints for integer/floating-point arithmetic

### LINKED ISSUE #887: Lint basic arithmetic on integers or floating point arithmetic
This would be an allow lint that could be activated for code where a more specific kind of arithmetic is wanted (e.g. wrapping or checked).

Courtesy of tomaka on reddit

### PR DESCRIPTION
**Pull request by llogiq**
This fixes #887 

I'm a bit unsure about % – can this overflow on integer?

### CODE REVIEW

**Comment by oli-obk** [comment]
Shouldn't we add a restrictions group?

% can overflow just like /

**Comment by mcarton** [comment]
`int_min {/,%} {0,-1}` are probably bad, the rest should be ok?

File: tests/compile-fail/arithmetic.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    i / 2; 
+    i - 2 + 2 - i; //~ERROR integer arithmetic detected
+    -i; //~ERROR integer arithmetic detected
+    
+    i & 1; // no wrapping
+    i | 1; 
+    i ^ 1;
+    i % 7;
+    i >> 1;
+    i << 1;
```
Comment (mcarton) [inline tests/compile-fail/arithmetic.rs]: Can’t `<<` overflow?

File: tests/compile-fail/arithmetic.rs [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#![feature(plugin)]
+#![plugin(clippy)]
+
+#![deny(integer_arithmetic, float_arithmetic)]
+#![allow(unused, shadow_reuse, shadow_unrelated, no_effect)]
+fn main() {
+    let i = 1i32;
+    1 + i; //~ERROR integer arithmetic detected
+    i * 2; //~ERROR integer arithmetic detected
+    1 % //~ERROR integer arithmetic detected
```
Comment (mcarton) [inline tests/compile-fail/arithmetic.rs]: You seem to be missing half an expression here.

File: tests/compile-fail/arithmetic.rs (hunk 1 above)
Comment (oli-obk) [inline tests/compile-fail/arithmetic.rs]: All possible shift overflow cases are already linted by rustc

Comment (mcarton) [inline tests/compile-fail/arithmetic.rs]: Ah, indeed, looks like it’s defined to be `255u8 << 1 == 254` and `255u8 * 2` ⇒ panic.

File: tests/compile-fail/arithmetic.rs (hunk 2 above)
Comment (llogiq) [inline tests/compile-fail/arithmetic.rs]: No, the expression is continued on the next line on purpose. This tests that only the outermost expression is linted.

Comment (mcarton) [inline tests/compile-fail/arithmetic.rs]: Ah ok. That, here, is confusing, maybe add a regular comment that explains that and/or put that in another section. Also `i / 2` should probably not lint, it does not panic ever.

**Comment by llogiq** [comment]
@oli-obk adding a restrictions group is a great idea. I think I'll add it to the PR once I find the time.

**Comment by llogiq** [comment]
Note to self: Also check for constants and don't lint if known good (e.g. `x / 2`).

**Comment by llogiq** [comment]
@mcarton r?

I think this is good enough to merge now, we can introduce improvements later.

* [ ] somehow identify restrictions lints in the `README.md`
* [ ] avoid linting e.g. division/remainder by known safe constant
* [ ] perhaps refactor the code a bit to reduce duplication

File: src/arithmetic.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+                    _ => ()
+                }
+                let (l_ty, r_ty) = (cx.tcx.expr_ty(l), cx.tcx.expr_ty(r));
+                if l_ty.is_integral() && r_ty.is_integral() {
+                    span_lint(cx,
+                              INTEGER_ARITHMETIC,
+                              expr.span,
+                              "integer arithmetic detected");
+                    self.span = Some(expr.span);
+                } else if l_ty.is_floating_point() && r_ty.is_floating_point() {
```
Comment (Manishearth) [inline src/arithmetic.rs]: if one is FP and one is integral we should count it as FP arithmetic.

File: util/update_lints.py [hunk 4]
```diff
@@ -21,12 +21,18 @@
     " (?P<desc>(?:[^"\\]+|\\.)*) " \s* [})]
 ''', re.VERBOSE | re.DOTALL)
 
+declare_restriction_lint_re = re.compile(r'''
+    declare_restriction_lint! \s* [{(] \s*
+    pub \s+ (?P<name>[A-Z_][A-Z_0-9]*) \s*,\s*
+    " (?P<desc>(?:[^"\\]+|\\.)*) " \s* [})]
```
Comment (Manishearth) [inline util/update_lints.py]: It is 3AM, so the regex-parsing section of my brain has gone to sleep -- I'm not even going to try :stuck_out_tongue_closed_eyes:

File: src/arithmetic.rs (hunk 3 above)
Comment (Manishearth) [inline src/arithmetic.rs]: apparently this is not possible. huh.

Comment (llogiq) [inline src/arithmetic.rs]: Not in my Rust :smile:

**Comment by Manishearth** [comment]
I'm going to assume the regex is sane and merge this.  We should next get the readme more restriction-aware, as you said, and update the wiki too.

File: src/arithmetic.rs (hunk 3 above)
Comment (mcarton) [inline src/arithmetic.rs]: It’s Rust, not C :smile:

File: util/update_lints.py (hunk 4 above)
Comment (llogiq) [inline util/update_lints.py]: It's basically copy&paste from the `declare_deprecated_lint` one.

File: src/arithmetic.rs (hunk 3 above)
Comment (Manishearth) [inline src/arithmetic.rs]: I've been hacking on gdb and some opengl project work lately, so still in that C/++ mindset :p

**Comment by llogiq** [comment]
cc @tomaka

File: util/update_lints.py (hunk 4 above)
Comment (mcarton) [inline util/update_lints.py]: At what time was that PR merged? :stuck_out_tongue:

**Pull request merged**